File: src/libs/SspUis.BizLogicLayer/Hrm/Info/ItemOfExpenseServices/Concrete/ItemOfExpenseService.cs
using GenericServices;
using StatusGeneric;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using WEBASE;
using WEBASE.AspNet;
using WEBASE.Models;
using SspUis.Core.Security;
using SspUis.DataLayer.EfClasses;
using SspUis.DataLayer;
using SspUis.DataLayer.Repositories;
using Microsoft.EntityFrameworkCore;
using SspUis.Core;
using SspUis.DataLayer.EfClasses.Hrm;

namespace SspUis.BizLogicLayer.Hrm.ItemOfExpenseServices
{
    public class ItemOfExpenseService : StatusGenericHandler, IItemOfExpenseService
    {
        private readonly IItemOfExpenseRepository _repository;
        private readonly IUnitOfWork _unitOfWork;
        private readonly IAuthService _authService;

        public ItemOfExpenseService(IUnitOfWork unitOfWork, IAuthService authService)
        {
            _repository = unitOfWork.ItemOfExpenseRepository;
            _unitOfWork = unitOfWork;
            _authService = authService;
        }

        public PagedResult<ItemOfExpenseListDto> GetList(SortFilterPageOptions dto)
        {
            var result = _repository.ReadAsNoTracked<ItemOfExpenseListDto>().SortFilter(dto).AsPagedResult(dto);
            return result;
        }

        public ItemOfExpenseDto Get()
        {
            return new ItemOfExpenseDto();
        }

        public ItemOfExpenseDto Get(int id)
        {
            var dto = _repository.ById<ItemOfExpenseDto>(id);
            CombineStatuses(_repository);
            return dto;
        }

        public SelectList<int> AsSelectList(int? itemOfExpenseId = null)
        {
            return _repository.AllAsQueryable
                .AsSelectList(itemOfExpenseId);
        }

        public HaveId<int> Create(CreateItemOfExpenseDlDto dto)
        {
            var entity = _repository.Create(dto, ent => Validation(dto, ent));
            CombineStatuses(_repository);
            if (IsValid)
            {
                _unitOfWork.Save();
                return HaveId.Create(entity.Id);
            }
            return null;
        }

        public void Update(UpdateItemOfExpenseDlDto dto)
        {
            _repository.Update(dto, ent => Validation(dto, ent));
            CombineStatuses(_repository);
            if (IsValid)
                _unitOfWork.Save();
        }

        public void Delete(int id)
        {
            try
            {
                _repository.Delete(id);
                CombineStatuses(_repository);
                if (IsValid)
                    _unitOfWork.Save();
            }
            catch (DbUpdateException)
            {
                AddError("Запись не может быть удален");
            }
        }

        private void Validation<TDto>(ItemOfExpenseDlDto<TDto> dto, ItemOfExpense entity)
            where TDto : ItemOfExpenseDlDto<TDto>
        {
            var query = _repository.AllAsQueryable;

            if (entity != null)
                query = query.Where(a => a.Id != entity.Id);

        }
    }
}

File: ssp_uis_frontend/spp_uis_frontend_erp/src/services/hrm/itemofexpense.service.js
import ApiService from '../api.service';

const ItemOfExpenseService = {
   GetList(data) {
      return ApiService.post('hrm/ItemOfExpense/GetList', data);
   },
   Get(id) {
      if (id == 0 || id === null || id === undefined) {
         return ApiService.get('hrm/ItemOfExpense/Get');
      } else {
         return ApiService.get(`hrm/ItemOfExpense/Get/${id}`);
      }
   },
   Update(data) {
      if (data.id == 0) {
         return ApiService.post(`hrm/ItemOfExpense/Create`, data);
      } else {
         return ApiService.post(`hrm/ItemOfExpense/Update`, data);
      }
   },
   Delete(id) {
      return ApiService.post(`hrm/ItemOfExpense/Delete/${id}`);
   },
   GetAsSelectList() {
      return ApiService.get(`hrm/ItemOfExpense/GetAsSelectList`);
   },
   Create(data) {
      if (data.id == 0) {
         return ApiService.post(`hrm/ItemOfExpense/CreateWithUser`, data);
      } else {
         return ApiService.post(`hrm/ItemOfExpense/UpdateWithUser`, data);
      }
   }
};
export default ItemOfExpenseService;
